magma/smart/smart_bits.py

```diff
@@ -0,0 +1,293 @@
+import abc
+import copy
+import enum
+import inspect
+import operator
+
+from magma.bits import Bits, BitsMeta
+from magma.conversions import uint, bits
+from magma.debug import debug_wire
+from magma.protocol_type import MagmaProtocolMeta, MagmaProtocol
+
+
+def _is_int(value):
+    try:
+        int(value)
+    except (ValueError, TypeError):
+        return False
+    return True
+
+
+class Context:
+    def __init__(self, assignee, root):
+        self.assignee = assignee
+        self.root = root
+
+    def max_width(self):
+
+        def _visit(node):
+            if hasattr(node, "_width_"):
+                return node._width_
+            if isinstance(node, _SmartBitsExpr):
+                return len(node.bits)
+            return max(_visit(arg) for arg in node.args)
+
+        max_width = _visit(self.root)
+        if self.assignee is not None:
+            max_width = max(max_width, len(self.assignee.bits))
+        return max_width
+
+
+class _SmartExprMeta(MagmaProtocolMeta):
+    pass
+
+
+class _SmartOpExprMeta(_SmartExprMeta):
+    pass
+
+
+class _SmartBitsExprMeta(_SmartExprMeta):
+    pass
+
+
+class Determination(enum.Enum):
+    SELF_DETERMINED = enum.auto()
+    CONTEXT_DETERMINED = enum.auto()
+
+
+class _SmartExpr(MagmaProtocol, metaclass=_SmartExprMeta):
+    def __init__(self, determination):
+        self._determination = determination
+
+    @abc.abstractmethod
+    def resolve(self):
+        raise NotImplementedError()
+
+    @property
+    def determination(self):
+        return self._determination
+
+    def __add__(self, other):
+        assert isinstance(other, _SmartExpr)
+        return _SmartBinaryOpExpr(operator.add, self, other)
+
+    def __le__(self, other):
+        assert isinstance(other, _SmartExpr)
+        return _SmartComparisonOpExpr(operator.le, self, other)
+
+    def __lshift__(self, other):
+        assert isinstance(other, _SmartExpr)
+        return _SmartShiftOpExpr(operator.lshift, self, other)
+
+    def __invert__(self):
+        return _SmartUnaryOpExpr(operator.invert, self)
+
+
+class _SmartOpExpr(_SmartExpr, metaclass=_SmartOpExprMeta):
+    def __init__(self, determination, op, *args):
+        super().__init__(determination)
+        self._op = op
+        self._args = tuple(args)
+
+    @property
+    def args(self):
+        return self._args
+
+    @property
+    def op(self):
+        return self._op
+
+    def _update(self, *args):
+        self._args = tuple(args)
+
+    def __str__(self):
+        if inspect.isbuiltin(self._op):
+            op_str = self._op.__name__
+        else:
+            op_str = str(self._op)
+        args = ", ".join(str(arg) for arg in self._args)
+        return f"{str(op_str)}({args})"
```

Comment:
Minor Do we need the extra str call on op_str here? Could either call str in the first if body to match the else, or remove the else from the else body?